### ISSUE TITLE
API call to clone an object

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#527]
It would be useful to have an API call to clone an object, i.e. create a new object and copy (some) properties over. There are many variants of cloning for different purposes for such an API call will probably need a set of control flags:

- Should the internal prototype be copied too, or should the new object inherit from Object.prototype?
- Should inherited properties be copied as own properties?
- Should only enumerated properties be copied, or all properties?
- What should be done about internal properties (Duktape specific)?
- What should be done about setters/getters - copy values or copy them as setters/getters?
- What should be done about property attributes?

The answers to these (and other questions) depend on what the clone primitive is used for so there isn't a single "clone" that will satisfy all use cases.

Feel free to discuss what you'd want clone :-)

**Comment by svaarala** [comment]
Some approaches existing code uses:

- `JSON.parse(JSON.stringify(x))`: shallow clone of only own properties
- [Object.assign()](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)
- [jQuery.extend()](https://api.jquery.com/jquery.extend/)
- [_.clone()](http://underscorejs.org/#clone)

These may be useful as inputs for useful/common semantics.

**Comment by fatcerberus** [comment]
Hm, this is a tricky question.  If JS had a built-in clone primitive, it would be easy to just replicate that functionality, but here's my input:

* For starters I think it's best if the clone is shallow, no recursive copying.  Deep cloning is opening up an entire high-level can of worms that I don't think Duktape need concern itself with.
* The internal prototype should be copied.  If `obj instanceof SomeCtor` is true, then `cloneObj instanceof SomeCtor` should be true also.
* Only own properties should be copied over directly.  Inherited properties will make it in through the shared prototype (see above).
* Internal own properties should be copied for the most part.  They are part of client code, just accessed from the native side instead of in script.
* Getters and setters should be translated over as such without calling the getter.
* Property attributes should be copied.

**Comment by svaarala** [comment]
One argument for supporting a shallow clone is that Duktape can do it more efficiently than the equivalent C or Ecmascript code. So why wouldn't the same argument hold for deep cloning?

But I agree deep cloning is much more complicated (DAGs and circular structures, etc), and I wouldn't want to go there at least initially :)

**Comment by svaarala** [comment]
@fatcerberus Your preferences would basically amount to a simple property table copy: all own properties would be copied as is. This would be quite efficient internally: one would just copy the array and entry parts used internally for property tracking, and bump the refcounts of keys/values by one.

**Comment by fatcerberus** [comment]
Regarding deep cloning, efficiency wasn't the issue in my mind--it's that, from what I've seen, deep copying is a complicated topic with many conflicting opinions on what the "right way" to do it is.  I don't think we'd want to go down that road with Duktape, best to leave it up to script.

**Comment by svaarala** [comment]
Yeah, that is definitely the case - it depends much more on context than shallow cloning.

But even for shallow cloning sometimes call sites now use `JSON.parse(JSON.stringify(x))` which will only clone *enumerable* own properties. Quite interestingly, it will clone inherited array elements (from Array.prototype) which almost no code actually uses but will nevertheless "see" inherited elements.

**Comment by svaarala** [comment]
Applying clone on a Proxy object is another interesting question :) Should it duplicate the Proxy *and* the Proxy target, or just the Proxy?

**Comment by fatcerberus** [comment]
Hm, that's an interesting question that I don't really have an answer for right now.  Maybe someone else could weigh in?

**Comment by svaarala** [comment]
Another option is actually that the clone result should be a bare object with all own properties copied by reading them through the Proxy. This is what would happen if one were to implement the clone in Ecmascript using `Object.getOwnPropertyNames()`. Property attributes etc are quite an interesting matter here too.

Thinking quickly I'd say the Proxy should be cloned as well as the target object; for the most part proxied objects behave like their bare counterparts so getting an independent copy of the target would maybe be expected. This doesn't guarantee the resulting object is independent of the original, though, because the Proxy traps may refer to private state (such as private backing object(s)) which won't get cloned.
